### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by balbasty**
Refactoring attempt. The idea is to separate the different types of modules a bit better:
* `core/` -> python or torch utilities that are used everywhere
* `spatial/` -> functional implementation of everything related to lattices and coordinate systems (interpolation, affine matrices, finite differences, etc.). In the best of worlds, everything should be automatically differentiable and callable from a `nn/` object. We're not quite there yet.
* `nn` -> neural networks stuff that relies on pytorch's Module class. In the best of worlds, everything has a differentiable functional implementation somewhere else (`core`, `spatial`, `vb`, ...) and is only wrapped here. We're not quite there yet.
* `vb/` -> probabilistic stuff. It's still not clear in my head what goes there. Long term, we'll have a lot of low-level probabilistic tools (likelihood, divergences, etc), plus mid-level fitting algorithms. Also, where do we store high-level, medical imaging stuff (say: registration models, or unified segmetnation)?
* `plots/` -> generic plotting utilities. We probably want to keep very specialized plotters next to their algorithms.
* `tools/` -> is this where we keep high-level stuff?

@brudfors this is the occasion to adapt your downstream code. The most painful part is that the content of `nitorch.mathfun` and `nitorch.utils` has been split up between multiple files in `core` (`pyutils`, `utils`, `math`, `linalg`).

Commits on `spatial._grid` in this pull request should be also applied in MONAI.

### CODE REVIEW

File: nitorch/_C/pushpull_common.cpp [hunk 1]
```diff
@@ -891,9 +891,9 @@ ::check3d(offset_t w, offset_t h, offset_t d, offset_t n) const
   scalar_t z = grid_ptr_NXYZ[grid_sC*2];
 
   // Check if out-of-bound
-  if (!(extrapolate || inbounds(x, src_X, static_cast<scalar_t>(TINY))
-                    || inbounds(y, src_Y, static_cast<scalar_t>(TINY))
-                    || inbounds(z, src_Z, static_cast<scalar_t>(TINY)))) {
+  if (!(extrapolate || (inbounds(x, src_X, static_cast<scalar_t>(TINY))
```
Comment (brudfors) [inline nitorch/_C/pushpull_common.cpp:894]: Are these fixes?

**Review by brudfors — COMMENTED** [review]
Thanks for dealing with UniRes for me! I am happy for you to merge into master at anytime. I hope to soon add a coreg module to tools.

File: nitorch/core/kernels.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 # . Use inplace operations if gradients not required
 
 import torch
-from nitorch import utils
+from ..core import utils
```
[the remark this answers is not in the document]
Comment (balbasty) [reply nitorch/core/kernels.py:12]: I thought we said that we would use relative imports?
https://github.com/balbasty/nitorch/issues/17

File: nitorch/core/constants.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+"""Useful constants."""
+
+import math
+import torch
+from .optionals import numpy as np
+
+pi = math.pi      # pi
```
[the remark this answers is not in the document]
Comment (balbasty) [reply nitorch/core/constants.py:7]: Is it a PEP thing I don't know about?
There are not in uppercase in python's `math` module so assumed that they didn't need to.

File: nitorch/_C/pushpull_common.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (balbasty) [reply nitorch/_C/pushpull_common.cpp:894]: Yes! I was not checking for out-of-bounds the right way, which made mode `extrapolate=False` not work correctly
(I was basically always extrapolating)

[the remark this answers is not in the document]
Comment (brudfors) [reply nitorch/_C/pushpull_common.cpp:894]: Ah! That is why setting extrapolate=False never worked for me... Sorry, should have told you ages ago about it..

File: nitorch/core/kernels.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brudfors) [reply nitorch/core/kernels.py:12]: Oh yea, I just got confused because I thought we had already changed to using that.

**Pull request merged**